### ISSUE TITLE
Probability of repeats should be a percentage of repeats. 

### ISSUE THREAD

**Issue by benedettafrida** [ODonoghueLab/versus#45]
the probability of repeats needs to be converted into a plain percentage. 
if I set the value at 0.5 it always need to repeat half of the questions, what needs to be random are the questions that we choose to repeat, not the number of repeats.

**Comment by boscoh** [comment]
> the probability of repeats needs to be converted into a plain percentage.

Is this referring to the units of probability? In most textbooks on probability that I've read, probabilities are written in the interval [0, 1]. All probability functions are expressed in that domain.

> if I set the value at 0.5 it always need to repeat half of the questions, what needs to be random are the questions that we choose to repeat, not the number of repeats.

To clarify, I use the probability to calculate the fraction of the questions that have to be repeated. The actual questions chosen to be repeated are chosen as the first item in a Fisher-Yates shuffle of the answers that have already been answered.

Each question that is to be repeated, will only be repeated once.

As well, I use the same probability to calculate the likelihood of showing a repeat as the next question, as opposed to the next question. Now for the case where one wants to test all questions P=1, there is no way to "intersperse" the repeat questions. For a fair distribution, it will be a case of Question-Repeat-Question-Repeat. However, if we do the case of Question-Question-Question-Repeat..., then you will pile up all the Repeats at the end .... Repeat-Repeat-Repeat - which of course, has no interspersing of the repeat questions.

Is this the way you want to do this?

**Comment by benedettafrida** [comment]
No, this is not the way the probability of repeats needs to be coded for the Rondo study version. 

Again, for the Rondo study, it does not make any sense to have a probability attached to how many questions we are going to have repeated.  It needs to be a  percentage.  With the Rondo study I will always know how many questions we are showing in total ( = the one I input), a different story is for the 2AFC. For example, if I want to have half of my questions repeated, the number of repeats should always be half of my input questions and shouldn't change between different runs. 

As I suggested  already a possible algorithm could be:

- Read the input value of “percentage of repeats”
- Calculate the overall number of repeated questions (cause you know how many I uploaded) 
- Randomly pick which questions are going to be repeated
- Create the list of questions (with the repeats) 
- Randomise all the questions order with a constraint that identical questions must be at least two questions apart


If it still not clear, let's schedule a skype meeting.

**Comment by boscoh** [comment]
I have checked the number of repeats, and have found it to be stable.

This is what I did:

- created an experiment "rondo-test" in the account user:"a@a.com";pw:"a". This is a replica of Benedetta's experiment "circle_2".
- set the "Fraction to be Repeated" to 0.3
![screen shot 2018-04-16 at 10 57 03 am](https://user-images.githubusercontent.com/180224/38788841-9b3fe1bc-4179-11e8-9b02-d3ec2b0c77d4.png)
- ran the experiment 5 times, choosing answers by clicking on 1 position:
![screen shot 2018-04-16 at 10 57 11 am](https://user-images.githubusercontent.com/180224/38788844-a1767dfc-4179-11e8-8466-acdb8b567364.png)
- I get the same number of repeats for all 5 experiments.

I also tested it on a version of Versus on my local server. 
- I use the same of images as "circle_2"
- I set the probability to 0.2
![screen shot 2018-04-16 at 1 26 26 pm](https://user-images.githubusercontent.com/180224/38788910-29ec1d0e-417a-11e8-85dd-f522717bf23c.png)
- I repeat the experiment 8 times:
![screen shot 2018-04-16 at 1 26 33 pm](https://user-images.githubusercontent.com/180224/38788930-40b08d04-417a-11e8-83a6-0a8612233b80.png)
- I get the same number of repeats 

In summary, I do not get a variable number of repeats for the same value of Fraction of Repeats.

Benedetta, can you confirm this on your account?

**Comment by benedettafrida** [comment]
Can confirm that the number of repeats is now consistent between runs. Thanks. 

**Question**: as an experiment, I set the fraction of repeats to 0.35 over a set of 15 questions. The number of repeats turns out to be 6 and not 5 like I would have expected (15*0.35 = 5.25 and to round to the closest integer should be 5). Do you know why?

**Comment by boscoh** [comment]
Typically, when converting Math.random for [0, 1) to indices for arrays, one does not use Math.round, but one chooses Math.ceil or Math.floor. I choose Math.ceil, so that you never get nRepeat= 0 if a Fraction of Repeat is set to non-zero.

**Comment by benedettafrida** [comment]
There are a couple of things that do not add up. First of all, I don't see why you're using Math.random to calculate the number of repeats (which I assume you call nRepeat). As we discussed multiple times there's no probability attached to calculating this number. Second, sure thing you get nRepeat=0 every time you set the fraction of repeats =0 (see the attached image). 

So, again: when calculating the number of repeats if I set fraction of repeats to 0.35 over a set of 15 questions I expect to see 5 questions repeated. Meaning nRepeats=Math.round(number_of_questions_I_input * fraction_of_repeats). 
![screen shot 2018-05-02 at 3 40 55 pm](https://user-images.githubusercontent.com/1155665/39507241-6d0230d0-4e20-11e8-8071-8a07749a50a9.png)

**Comment by boscoh** [comment]
1. I do not use Math.random in your section of the code
2. In the screenshot, you've set the Fraction of Repeat = 0, therefore, if you use that setting, there will be 0 repeat
3. Currently, I've allowed the Fraction of Repeat to be changed, however, this means the number of repeats in the sample set will depend on the value of Fraction of Repeat at the time of completion of the experiment.

**Comment by benedettafrida** [comment]
Sorry, maybe it wasn't clear that I was replying to your previous comment 

> Typically, when converting Math.random for [0, 1) to indices for arrays, one does not use Math.round, but one chooses Math.ceil or Math.floor. I choose Math.ceil, so that you never get nRepeat= 0 if a Fraction of Repeat is set to non-zero.

Which should clarify your point 1 and 2. 

>3.  Currently, I've allowed the Fraction of Repeat to be changed, however, this means the number of repeats in the sample set will depend on the value of Fraction of Repeat at the time of completion of the experiment.

What do you mean by it? I don't understand what you mean by  

>  will depend on the value of Fraction of Repeat at the time of completion of the experiment

What I need is very simple: when I set the fraction of repeats to 0.35 over a set of 15 questions the number of repeats turns out to be 6 (see screenshot before 2nd experiment) and not 5 like I would have expected (15*0.35 = 5.25 and to round to the closest integer should be 5).

**Comment by boscoh** [comment]
For some reason, you've set the Fraction of Repeats to be 0 in the screen-shot.

**Comment by boscoh** [comment]
It is not 0.35

**Comment by benedettafrida** [comment]
![screen shot 2018-05-02 at 3 38 54 pm](https://user-images.githubusercontent.com/1155665/39508807-189e5288-4e27-11e8-894c-43e5765da46d.png)

**Comment by benedettafrida** [comment]
I hope it's clearer now.

**Comment by boscoh** [comment]
Actually, it should be `Math.floor(Math.random() * (max - min)) + min` from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/random.

I used Math.ceil, which is wrong.

Math.floor should be used is pretty well explained here: https://stackoverflow.com/a/7377769

**Comment by benedettafrida** [comment]
Honestly, I don't understand why are you trying to get a random number between two values. Again we shouldn't use any random values for this calculation. 
To repeat, this is what I want in pseudocode:

```
number_Repeats = total_Number_Of_Questions * fraction_of_repeats
x = numbers AFTER the decimal point of the number_Repeats
y= number BEFORE the decimal point of the number_Repeats

if(x > 0 && x < 49 ){
 number_Repeats = y;
} else if (x >= 50) {
 number_Repeats = y+1;
}
```


This means, if number_Repeats is 5.25, then this number should be rounded to the closest smallest integer (i.e. 5). If number_Repeats is 5.55, then this number should be rounded up to the closest largest integer (i.e. 6). So in the case of the screenshot, with the total number of questions being 15, and the fraction of repeats being 0.35, number_Repeats is 5.25, x is 25, y=5,  which means, it should be rounded to 5.  
As explained here:  https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/round

**Comment by boscoh** [comment]
The problem I am alluding is the mapping of an interval [0, 1) to an interval [m,n] where m,n is integral. We want the mapping to be uniformly distributed over the integers from m to n. I understand that you might be confused to see it in the context of Math.random, but discussions in the literature refer to Math.random() because it is considered the canonical function to generate an uniform distribution between [0, 1). It is the inteval that matters. In our case, we enter values by hand that are in this interval.

Your algorithm using round is generally considered to be an uneven mapping between [0, 1) to [m, n]. It is non-uniformly distributed. It is not mathematically sound. Still, I will use it if you that is what you want.

**Comment by benedettafrida** [comment]
Honestly, this is rather frustrating. We discussed several times that the number of repeats, meaning the NUMBER of the questions that overall should berepeated has to be consistent between different runs of the same experiment, therefore shouldn't use random numbers of any distribution. I don't understand why you keep referring to random numbers when I really thought we weren't using them anymore to calculate the NUMBER of repeats. 

All I am saying is that if I set set the fraction of repeats to 0.35 over a set of 15 questions the number of repeats turns out to be 6 (see screenshot before 2nd experiment) and not 5 like I would have expected (15*0.35 = 5.25 and to round to the closest integer should be 5). See the screenshot. 

![screen shot 2018-05-02 at 3 38 54 pm](https://user-images.githubusercontent.com/1155665/39512146-2c2a4ca6-4e33-11e8-961a-ec2842fe091c.png)

**Comment by boscoh** [comment]
Yes, agreed. This is frustrating.

Basically, Math.round is rarely used in any serious code I've seen, and Math.ceil is generally used.

I see now that the reasons are very difficult to explain without a solid background in applied-math/statistics and/or algorithms. Apologies if my explanations seem to go around in circles.

Nevertheless, for the case here, the differences between the two are tiny and unlikely to cause problems. I will use Math.round if that is what you are comfortable with.

**Comment by benedettafrida** [comment]
In "serious codes"  Math.round, Math.ceil and Math.floor, Math.random (as well as any other function really), are applied when they need to be applied. None of them is universally used, it depends on what you're trying to achieve. 

Here what we are trying to achieve is to get an integer from the multiplication of two values that a user will input (number of total question and fraction). More specifically, an integer that a user will likely to consider correct. 
If I were to ask you, without the need to code anything, what's the closest integer to 5.2? I'm pretty sure you would say 5 and not 6 (because I am assuming you have some basic mathematical knowledge).   

So I guess that the real reason of this entire waste of time is your lack of understanding of what this algorithm is supposed to do and why. Do we need to use random functions to determine this number? Absolutely not, it doesn't make any sense. Do we need to select integers according to a uniform distribution (or any other type of distribution for that matter) for THIS value? Again no, it doesn't make any sense. 

I hope this can be fixed without any more waste of time.

**Comment by boscoh** [comment]
Apologies for the tone of my last post, I didn't mean to cause offence. I will do as you want.
